pkg/detectors/uri/uri.go

```diff
@@ -23,7 +23,7 @@ var _ detectors.Detector = (*Scanner)(nil)
 var _ detectors.CustomFalsePositiveChecker = (*Scanner)(nil)
 
 var (
-	keyPat = regexp.MustCompile(`\b(?:https?:)?\/\/[\S]{3,50}:([\S]{3,50})@[-.%\w\/:]+\b`)
+	keyPat = regexp.MustCompile(`\b(?:https?:)?\/\/[\w-\.]{3,50}:([\w-\.%$^&#@]{3,50})@[-.%\w\/:]+\b`)
```

Comment:
A good source of truth is [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986#section-3.2.1). An unencoded `@` isn't a valid password character because it's the delimiter between `userinfo` and `host`; likewise, `:` isn't a valid username character.

The right-hand side of `@` is also problematic. e.g., [`%` and `/` aren't valid characters for a host](https://datatracker.ietf.org/doc/html/rfc3986#section-3.2.2).

Reply:
I improved the regex and removed the invalid chars as you mentioned.